truckersmp-cli

```diff
@@ -1,512 +1,596 @@
-            || error=1
-        fi
-    done
-
-    # something went wrong
-    if [ "$error" = "1" ]; then
-        >&2 echo "Failed to download mod files."
-        quit 1
-    fi
-}
-
-update_game() {
+    modfiles = []
+    try:
+        for item in JSONDecoder().decode(str(files_json, "ascii"))["Files"]:
+            modfiles.append((item["Md5"], item["FilePath"]))
+        if len(modfiles) == 0:
+            raise Exception("File list is empty")
+    except Exception as e:
+        sys.exit("""Failed to parse files.json: {}
+Please report an issue: {}""".format(e, URL.issueurl))
+
+    # compare existing local files with md5sums
+    # and remember missing/wrong files
+    dlfiles = []
+    for md5, jsonfilepath in modfiles:
+        md5hash = hashlib.md5()
+        modfilepath = os.path.join(args.moddir, jsonfilepath[1:])
+        if not os.path.isfile(modfilepath):
+            dlfiles.append(("/files" + jsonfilepath, modfilepath, md5))
+        else:
+            try:
+                with open(modfilepath, "rb") as f:
+                    while True:
+                        buf = f.read(md5hash.block_size * 4096)
+                        if not buf:
+                            break
+                        md5hash.update(buf)
+                if md5hash.hexdigest() != md5:
+                    dlfiles.append(("/files" + jsonfilepath, modfilepath, md5))
+            except Exception as e:
+                sys.exit("Failed to read {}: {}".format(modfilepath, e))
+    if len(dlfiles) > 0:
+        printdbg("Files to download:")
+        for path, _, _ in dlfiles:
+            printdbg("  {}".format(path))
+    else:
+        printdbg("No files to download")
+
+    # download missing/wrong files
+    if not download_files(URL.dlurl, dlfiles):
+        if not download_files(URL.dlurlalt, dlfiles):
+            # something went wrong
+            sys.exit("Failed to download mod files.")
+
+
+def update_game():
     # make sure steam is closed before updating
     # it's possible to update with the steam client open but the client looses
     # all connectivity and asks for password and steam guard code after restart
-    while pidof steam > /dev/null 2>&1; do
-        print "found running steam, running pkill -x steam"
-        pkill -x steam
-    sleep 5
-    done
-
-    if ! [ -d "$gamedir" ]; then
-        print "mkdir $gamedir"
-        mkdir -p "$gamedir"
-    fi
+    try:
+        # newline-separated list of steam pids
+        steam_pids_str = str(
+          subproc.check_output(["pgrep", "-u", getuser(), "-x", "steam"]), "ascii")
+        # list of steam pid strings
+        steam_pid_list = steam_pids_str.splitlines()
+    except Exception:
+        pass
+    else:
+        printdbg("Found running steam, killing the processes")
+        # wait until all processes are killed
+        while len(steam_pid_list) > 0:
+            pid = int(steam_pid_list[0])
+            try:
+                os.kill(pid, signal.SIGTERM)
```

Comment:
I just found out about `steam -shutdown` maybe that's more suitable to quit the process.

Reply:
Does this include a fix for this? https://github.com/lhark/truckersmp-cli/issues/34

Reply:
This PR is not intended to fix it (but this PR fixes #32).